File: leetcode/15. 3Sum/solution1.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        
        const int nums_length = nums.size();
        
        sort(nums.begin(), nums.end());
        
        vector <vector <int>> ret;
        
        if (nums.size() < 3) {
            return ret;
        }
        
        int prev_i = INT_MIN;
        
        for (int i = 0; i < nums_length; ++i) {
            
            if (prev_i == nums[i]) {
                continue;
            }
            
            int l = i + 1, r = nums_length - 1;
            int ele = nums[i];
            int prev_l = INT_MIN;
            int prev_r = INT_MIN;
            
            while (l < r) {
                
                if (nums[i] + nums[l] + nums[r] == 0) {
                    ret.push_back({nums[i], nums[l], nums[r]});

                    prev_l = nums[l];
                    prev_r = nums[r];

                    
                    while (nums[l] == prev_l and nums[r] == prev_r and l < r) {
                        r--;
                        l++;
                    }
                }
                
                else if (nums[i] + nums[l] + nums[r] > 0) {
                    r--;
                }
                else {
                    l++;
                }
            }
            
            prev_i = nums[i];
        }
        
        return ret;
        
    }
};

File: leetcode/15. 3Sum/solution1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        
        if len(nums) < 3:
            return []
        
        nums.sort()
        
        curr_master_ele = None
        prev_j_ele = None
        prev_k_ele = None
        solution_set = []
        
        for i in range(len(nums) - 2):
            if curr_master_ele != nums[i]:
                curr_master_ele = nums[i]
            
            else:
                continue
            
            j = i + 1
            k = len(nums) - 1
            
            while j < k:
                curr_sum = curr_master_ele + nums[j] + nums[k]
                
                if curr_sum == 0:
                    
                    if prev_j_ele == nums[j] and prev_k_ele == nums[k]:
                        j += 1
                        k -= 1
                        
                        continue
                    
                    solution_set.append([curr_master_ele,
                                        nums[j],
                                        nums[k]])
                    
                    prev_j_ele = nums[j]
                    prev_k_ele = nums[k]
                    
                    j += 1
                    k -= 1
                
                elif curr_sum > 0:
                    k -= 1
                
                else:
                    j += 1
            
        return solution_set
